I run a test with this change on my ubuntu 18.04 with a 2080Ti GPU:
```
from tensorflow.keras.layers import Input, Embedding, Bidirectional, GRU, Dense
from tensorflow.keras.models import Model
from transformers import TFDistilBertModel
from tensorflow.keras.mixed_precision import experimental as mixed_precision
policy = mixed_precision.Policy('mixed_float16')
mixed_precision.set_policy(policy)

bert = TFDistilBertModel.from_pretrained('distilbert-base-uncased')
inputs = Input(shape=(None,), dtype='int32')
bert_out = bert(inputs)[0]
output = Dense(9, activation='softmax', dtype='float32')(bert_out)
model = Model(inputs, output)
model.compile(loss='sparse_categorical_crossentropy', optimizer='adam', metrics=['accuracy'])
model.summary()
x = [[5, 2, 3] * 3] * 100
y = [[1, 2, 3] * 3] * 100
model.fit(x=x, y=y, epochs=20, batch_size=16)
```
And get error info:
```
Traceback (most recent call last):
  File "test.py", line 8, in <module>
    bert = TFDistilBertModel.from_pretrained('distilbert-base-uncased')
  File "/home/xingya/transformers/src/transformers/modeling_tf_utils.py", line 602, in from_pretrained
    model(model.dummy_inputs, training=False)  # build the network with dummy inputs
  File "/home/xingya/.conda/envs/transformers/lib/python3.8/site-packages/tensorflow/python/keras/engine/base_layer.py", line 968, in __call__
    outputs = self.call(cast_inputs, *args, **kwargs)
  File "/home/xingya/transformers/src/transformers/modeling_tf_distilbert.py", line 615, in call
    outputs = self.distilbert(inputs, **kwargs)
  File "/home/xingya/.conda/envs/transformers/lib/python3.8/site-packages/tensorflow/python/keras/engine/base_layer.py", line 968, in __call__
    outputs = self.call(cast_inputs, *args, **kwargs)
  File "/home/xingya/transformers/src/transformers/modeling_tf_distilbert.py", line 508, in call
    tfmr_output = self.transformer(
  File "/home/xingya/.conda/envs/transformers/lib/python3.8/site-packages/tensorflow/python/keras/engine/base_layer.py", line 968, in __call__
    outputs = self.call(cast_inputs, *args, **kwargs)
  File "/home/xingya/transformers/src/transformers/modeling_tf_distilbert.py", line 401, in call
    layer_outputs = layer_module(hidden_state, attn_mask, head_mask[i], output_attentions, training=training)
  File "/home/xingya/.conda/envs/transformers/lib/python3.8/site-packages/tensorflow/python/keras/engine/base_layer.py", line 968, in __call__
    outputs = self.call(cast_inputs, *args, **kwargs)
  File "/home/xingya/transformers/src/transformers/modeling_tf_distilbert.py", line 355, in call
    ffn_output = self.ffn(sa_output, training=training)  # (bs, seq_length, dim)
  File "/home/xingya/.conda/envs/transformers/lib/python3.8/site-packages/tensorflow/python/keras/engine/base_layer.py", line 968, in __call__
    outputs = self.call(cast_inputs, *args, **kwargs)
  File "/home/xingya/transformers/src/transformers/modeling_tf_distilbert.py", line 304, in call
    x = self.activation(x)
  File "/home/xingya/.conda/envs/transformers/lib/python3.8/site-packages/tensorflow/python/keras/engine/base_layer.py", line 968, in __call__
    outputs = self.call(cast_inputs, *args, **kwargs)
  File "/home/xingya/.conda/envs/transformers/lib/python3.8/site-packages/tensorflow/python/keras/layers/core.py", line 420, in call
    return self.activation(inputs)
  File "/home/xingya/transformers/src/transformers/modeling_tf_distilbert.py", line 79, in gelu
    cdf = 0.5 * (1.0 + tf.math.erf(x / tf.math.sqrt(2.0)))
  File "/home/xingya/.conda/envs/transformers/lib/python3.8/site-packages/tensorflow/python/ops/math_ops.py", line 984, in binary_op_wrapper
    return func(x, y, name=name)
  File "/home/xingya/.conda/envs/transformers/lib/python3.8/site-packages/tensorflow/python/ops/math_ops.py", line 1081, in _truediv_python3
    raise TypeError("x and y must have the same dtype, got %r != %r" %
TypeError: x and y must have the same dtype, got tf.float16 != tf.float32
```
I made a modification to L299:
`self.activation = (
            tf.keras.layers.Activation(gelu, dtype='float32') if config.activation == "gelu" else tf.keras.activations.relu
        )`
And then the model began to train, however the loss don't decrease and the accuracy is always 0:
```
7/7 [==============================] - 0s 28ms/step - loss: 2.1972 - accuracy: 0.0000e+00
Epoch 2/20
7/7 [==============================] - 0s 29ms/step - loss: 2.1972 - accuracy: 0.0000e+00
Epoch 3/20
7/7 [==============================] - 0s 30ms/step - loss: 2.1972 - accuracy: 0.0000e+00
Epoch 4/20
7/7 [==============================] - 0s 31ms/step - loss: 2.1972 - accuracy: 0.0000e+00
```

I have trid this code in float32 precision, and it works. 
```
Epoch 1/20
7/7 [==============================] - 0s 31ms/step - loss: 2.5418 - accuracy: 0.2800
Epoch 2/20
7/7 [==============================] - 0s 33ms/step - loss: 1.2452 - accuracy: 0.3356
Epoch 3/20
7/7 [==============================] - 0s 31ms/step - loss: 1.1438 - accuracy: 0.3267
Epoch 4/20
7/7 [==============================] - 0s 33ms/step - loss: 1.1219 - accuracy: 0.3400
```

--- a/src/transformers/modeling_tf_distilbert.py
+++ b/src/transformers/modeling_tf_distilbert.py
@@ -76,7 +76,7 @@ def gelu(x):
         0.5 * x * (1 + torch.tanh(math.sqrt(2 / math.pi) * (x + 0.044715 * torch.pow(x, 3))))
         Also see https://arxiv.org/abs/1606.08415
     """
-    cdf = 0.5 * (1.0 + tf.math.erf(x / tf.math.sqrt(2.0)))
+    cdf = 0.5 * (1.0 + tf.math.erf(x / tf.cast(tf.math.sqrt(2.0), dtype=x.dtype)))
     return x * cdf
 
 

--- a/src/transformers/modeling_tf_distilbert.py
+++ b/src/transformers/modeling_tf_distilbert.py
@@ -263,9 +263,12 @@ class TFMultiHeadSelfAttention(tf.keras.layers.Layer):
         scores = tf.matmul(q, k, transpose_b=True)  # (bs, n_heads, q_length, k_length)
         mask = tf.reshape(mask, mask_reshape)  # (bs, n_heads, qlen, klen)
         # scores.masked_fill_(mask, -float('inf'))            # (bs, n_heads, q_length, k_length)
-        scores = scores - 1e30 * (1.0 - mask)
 
-        weights = tf.nn.softmax(scores, axis=-1)  # (bs, n_heads, qlen, klen)
+        scores_dtype = scores.dtype
+        # calculate `scores` in `tf.float32` to avoid numeric overflow
+        scores = tf.cast(scores, dtype=tf.float32) - 1e30 * (1.0 - tf.cast(mask, dtype=tf.float32))
+
+        weights = tf.cast(tf.nn.softmax(scores, axis=-1), dtype=scores_dtype)  # (bs, n_heads, qlen, klen)
         weights = self.dropout(weights, training=training)  # (bs, n_heads, qlen, klen)
 
         # Mask heads if we want to

